sv.c

```diff
@@ -4771,6 +4774,429 @@ Perl_sv_setsv_flags(pTHX_ SV *dsv, SV* ssv, const I32 flags)
+
+    const U32 sflags = SvFLAGS(ssv);
+    const STRLEN cur = SvCUR(ssv);
+    const STRLEN len = SvLEN(ssv);
+
+    assert(!SvIsCOW_static(ssv)); /* SVppv_STATIC: see S_sv_freshcopy_flags */
+
+    if (!(flags & SV_NOSTEAL) && S_SvPV_can_swipe_buf(ssv, sflags, cur, len) ) {
+        /* Passes the swipe test.  */
+        SvLEN_set(dsv, len);
+        SvCUR_set(dsv, cur);
+        SvPV_set(dsv, SvPVX_mutable(ssv));
+
+        assert(!SvOOK(ssv)); /* According to S_SvPV_can_swipe_buf() */
+        /* (void)SvOK_off(ssv); but without the superfluous SvOOK_off(ssv)) */
+        SvFLAGS(ssv) &= ~(SVf_OK|SVf_IVisUV|SVf_UTF8|SVs_TEMP);
+
+        SvPV_set(ssv, NULL);
+        SvLEN_set(ssv, 0);
+        SvCUR_set(ssv, 0);
+        return;
+    }
+
+    ASSUME(!(SvFLAGS(dsv) & SVf_BREAK));
+    /* S_SvPV_shared_hkey_or_CoWable() needs an accurate SvLEN(dsv) */
+    SvLEN_set(dsv, 0);
+
+    if ((flags & SV_COW_SHARED_HASH_KEYS) &&
+        S_SvPV_shared_hkey_or_CoWable(ssv, dsv, sflags, cur, len)
+    ) {
+        /* Either it's a shared hash key, or it's suitable for
+           copy-on-write.  */
+#ifdef DEBUGGING
+        if (DEBUG_C_TEST) {
+            PerlIO_printf(Perl_debug_log, "Copy on write: ssv --> dsv\n");
+            sv_dump(ssv);
+            sv_dump(dsv);
+        }
+#endif
+#ifdef PERL_ANY_COW
+        if (!(sflags & SVf_IsCOW)) {
+                SvIsCOW_on(ssv);
+                CowREFCNT(ssv) = 0;
+        }
+
+        if (len) {
+           if (sflags & SVf_IsCOW) {
+                sv_buf_to_rw(ssv);
+            }
+            CowREFCNT(ssv)++;
+            SvPV_set(dsv, SvPVX_mutable(ssv));
+            sv_buf_to_ro(ssv);
+        } else
+#endif
+        {
+        /* SvIsCOW_shared_hash */
+        DEBUG_C(PerlIO_printf(Perl_debug_log,
+                              "Copy on write: Sharing hash\n"));
+        SvPV_set(dsv,
+                 HEK_KEY(share_hek_hek(SvSHARED_HEK_FROM_PV(SvPVX_const(ssv)))));
+        }
+        SvLEN_set(dsv, len);
+        SvIsCOW_on(dsv);
+    } else {
+        /* Failed the swipe test, and we cannot do copy-on-write either.
+           Have to copy the string.  */
+
+#ifdef DEBUGGING
+        /* Perl_sv_grow_fresh asserts that cur == 0
+         * but doesn't actually need it to. */
+        SvCUR_set(dsv, 0);
+#endif
+        sv_grow_fresh(dsv, cur + 1);
+        Move(SvPVX_const(ssv),SvPVX(dsv),cur,char);
+        *(dsv->sv_u.svu_pv + cur) = '\0';
+    }
+    SvCUR_set(dsv, cur);
+    return;
+}
+
+/* S_sv_freshcopy_PVxx mainly handles source SV types above SVt_PV
+ * for Perl_sv_freshcopy_flags. This function may get inlined, even
+ * though it might be preferable if it didn't.
+ *
+ * Notes: it is the caller's responsibility to check GET magic.
+ *        Perl_sv_setsv_flags essentially ignores magic, except for
+ *        taint and vstring magic, which are also handled here. */
+
+static void
+S_sv_freshcopy_PVxx(pTHX_ SV* dsv, SV* ssv, const I32 flags)
```

Comment:
Change retval type to `SV*` and return  `SV* dsv` as the retval. Never waste a free register. Tailcall or reducing liveness or reducing peak non volatile reg usage opportunities exist now, or will exist in the future.

Reply:
I did try it and (IIRC) it performed either worse or no better. I'll look again before force-pushing changes for some of the other comments above though.